internal/v2/fakes2av2/fakes2av2.go

```diff
@@ -112,3 +135,116 @@ func getTlsConfiguration(req *s2av2pb.GetTlsConfigurationReq) (*s2av2pb.SessionR
+	}
+	if _, err := x509LeafCert.Verify(opts); err != nil {
+		s := fmt.Sprintf("Client Peer Verification failed: %v", err)
+		return buildValidatePeerCertificateChainSessionResp(3, s, s2av2pb.ValidatePeerCertificateChainResp_FAILURE, s, &s2av2ctx.S2AContext{}), err
+	}
+	return buildValidatePeerCertificateChainSessionResp(0, "", s2av2pb.ValidatePeerCertificateChainResp_SUCCESS, "Client Peer Verification succeeded", &s2av2ctx.S2AContext{
+		LeafCertSpiffeId: "", // TODO(rmehta19)
+		LeafCertUris: x509LeafCert.PermittedURIDomains,
+		LeafCertDnsnames: x509LeafCert.PermittedDNSDomains,
+		PeerCertificateChainFingerprints: []string{""}, // TODO(rmehta19)
+		LocalLeafCertFingerprint: x509LeafCert.Signature,
+	}), nil
+}
+
+func verifyServerPeer(req *s2av2pb.ValidatePeerCertificateChainReq) (*s2av2pb.SessionResp, error) {
+	derCertChain := req.GetServerPeer().CertificateChain
+	if len(derCertChain) == 0 {
+		err := errors.New("Server Peer Verification failed: server cert chain is empty.")
+		return buildValidatePeerCertificateChainSessionResp(3, err.Error(), s2av2pb.ValidatePeerCertificateChainResp_FAILURE, err.Error(), &s2av2ctx.S2AContext{}), err
+	}
+
+	// Obtain the set of root certificates.
+	rootCertPool := x509.NewCertPool()
+	if ok := rootCertPool.AppendCertsFromPEM(serverCert); ok != true {
+		err := errors.New("Server Peer Verification failed: S2Av2 could not obtain/parse roots.")
+		return buildValidatePeerCertificateChainSessionResp(13, err.Error(), s2av2pb.ValidatePeerCertificateChainResp_FAILURE, err.Error(), &s2av2ctx.S2AContext{}), err
+	}
+
+	// Set the Intermediates: certs between leaf and root, excluding the leaf and root.
+	intermediateCertPool := x509.NewCertPool()
+	for i := 1; i < (len(derCertChain) - 1); i++ {
+		x509Cert, err := x509.ParseCertificate(derCertChain[i])
+		if err != nil {
+			return buildValidatePeerCertificateChainSessionResp(3, err.Error(), s2av2pb.ValidatePeerCertificateChainResp_FAILURE, err.Error(), &s2av2ctx.S2AContext{}), err
+		}
+		intermediateCertPool.AddCert(x509Cert)
+	}
+
+	// Verify the leaf certificate.
+	opts := x509.VerifyOptions{
+		CurrentTime: time.Now(),
+		Roots: rootCertPool,
+		Intermediates: intermediateCertPool,
+	}
+	x509LeafCert, err := x509.ParseCertificate(derCertChain[0])
+	if err != nil {
+		s := fmt.Sprintf("Server Peer Verification failed: %v", err)
+		return buildValidatePeerCertificateChainSessionResp(3, s, s2av2pb.ValidatePeerCertificateChainResp_FAILURE, s, &s2av2ctx.S2AContext{}), err
+	}
+	if _, err := x509LeafCert.Verify(opts); err != nil {
+		s := fmt.Sprintf("Server Peer Verification failed: %v", err)
+		return buildValidatePeerCertificateChainSessionResp(3, s, s2av2pb.ValidatePeerCertificateChainResp_FAILURE, s, &s2av2ctx.S2AContext{}), err
+	}
+
+	return buildValidatePeerCertificateChainSessionResp(0, "", s2av2pb.ValidatePeerCertificateChainResp_SUCCESS, "Server Peer Verification succeeded",&s2av2ctx.S2AContext{
+		LeafCertSpiffeId: "", // TODO(rmehta19)
```

Comment:
Is it ok if we send an empty context for now and come back and populate it later?

Rationale: We need to populate all of this data based on your test certs, but your test certs probably don't have a SPIFFE ID, URI or DNS SANs, so we may need to change them.

Similarly for the server-side context.

Reply:
Will do, thanks for the rationale!